[eventpipe] Perform endian aware serialization of structs and arrays above FireETW layer

stefan-sf-ibm opened the pull request:
In the past struct members of payloads were always of primitive types. Since the enablement of the event `BulkType` in commit bfc7359d7e9 struct members may be of array type. On big-endian architectures we need to walk over the whole structure and perform byte swaps accordingly which is done by this fix also for array members.

According to the [manifest](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/ClrEtwAll.man) there is no event which makes use of nested structures or nested arrays except what is fixed by this patch. Assuming this won't change in the future this patch should be sufficient. Otherwise we might need a more general approach in order to parse and convert arbitrary payloads.

This is basically a follow up to #68648.

stefan-sf-ibm commented:
//CC @lateralusX

lateralusX commented:
Thanks for fixing this, will look at the PR in coming days.

lateralusX commented:
@stefan-sf-ibm So if I read the diff correctly, the change will only apply to BIGENDIAN platforms, correct? All other platforms won't be affected at all?

lateralusX commented on src/coreclr/scripts/genEventPipe.py:
```diff
@@ -284,28 +284,31 @@ def winTypeToFixedWidthType(t):
-                        pack_list.append("        valuePtr += value_len + 1;")
-                    elif t == "UTF16String":
-                        pack_list.append("        value_len = strlen((const char *)valuePtr);")
-                        pack_list.append("        success &= write_buffer_string_utf8_to_utf16_t((const ep_char8_t *)valuePtr, value_len, &buffer, &offset, &size, &fixedBuffer);")
```
This did a utf8 to utf16 conversion, and wrote the converted data into the buffer. It is now replaced by a call to write_buffer_string_utf16, but that assumes input is in utf16 since there is no utf8 -> utf16 conversion, something missing?

stefan-sf-ibm replied:
My previous understanding was that the source string coming from Mono was UTF-8 encoded. Though, this does not seem to be true. For the BulkType event the source string is already in UTF-16 format (although at the moment only an empty string is emitted). Thus I'm only left with byte swaps for UTF-16 strings.

https://github.com/dotnet/runtime/blob/a34291586e5f82a98f6b5a2e9134cee96f7fc306/src/mono/mono/eventpipe/ep-rt-mono.c#L3187

lateralusX replied:
OK, probably a side effect of how the bulk type event is serialized, normally we pass strings as UTF8 down to FireETW layer where the conversion and serialization takes place directly into the buffer, but for the bulk type events it gets serialized above the FireETW layer, ep_rt_mono_fire_bulk_type_event on Mono and BulkTypeEventLogger::FireBulkTypeEven on CoreCLR. On Mono we represent unicode strings as UTF8 and then on serialize layer we do the conversion, UTF8->UTF16, but as said since the serialization layer for this event type is partly moved up on level causing issues. When we serialize structs, we normally use the the ep_write_buffer_x that would already do byte swapping when serializing into the buffer, so maybe it would be better to fix this when where we serialize the structs instead of after in the serialized byte stream as done below? Sounds like this is mainly an issue in how bulk type event is serialized and that it should use existing ep_write_buffer_x instead of its own versions.

stefan-sf-ibm replied:
Thanks for the clarification. I will come up with an implementation which makes use of `write_buffer_x` in the FireETW layer for the bulk type event.

lateralusX replied:
What I think we should do is drop all the conversion done in the FireETW layer for structs that are part of a byte buffera and fix ep_rt_mono_fire_bulk_type_event to use our ep_write_buffer_x versions instead of its own buffer serializing functions unaware of endieness. The existing structs should already be written into the byte arrays using the ep_write_buffer_x so shouldn't need any additional work and could be serialized as is.

lateralusX commented on src/coreclr/scripts/genEventPipe.py:
```diff
@@ -618,6 +621,18 @@ def getMonoEventPipeHelperFileImplPrefix():
     size_t *size,
     bool *fixed_buffer);
 
+bool
```
Will these new methods only be called when BIGENDIAN is defined?

stefan-sf-ibm replied:
Yes, actually those are only called for big-endian archs, though, I implemented them endian independent just in case.

lateralusX commented on src/coreclr/scripts/genEventPipe.py:
```diff
@@ -731,6 +746,48 @@ def getMonoEventPipeHelperFileImplPrefix():
+}
+
+size_t
+strlen_utf16_t (const uint8_t *value)
```
ep_rt_utf16_string_len ?

stefan-sf-ibm replied:
Function `ep_rt_utf16_string_len` expects a pointer to an `ep_char16_t` which is 2 byte aligned and I'm coming in with a potentially 1 byte aligned pointer. Therefore, I came up with function `strlen_utf16_t`.

stefan-sf-ibm commented:
Yes, the patch affects only big-endian platforms.

stefan-sf-ibm commented:
@lateralusX ping

lateralusX commented:
> @lateralusX ping

Sorry for delay, I added a comment that maybe we should fix this at the source, at bulk type event since it looks like it really should use the ep_write_buffer_x instead, making sure that any byte arrays send down to FireETW layer already have right byte encoding, that is how other structs currently works.

stefan-sf-ibm commented:
> I added a comment that maybe we should fix this at the source, at bulk type event since it looks like it really should use the ep_write_buffer_x instead, making sure that any byte arrays send down to FireETW layer already have right byte encoding for all its content, that is how other structs currently works, like JITDone. Speaking of that, so it uses ep_write_buffer_uint64_t that does a byte swap on big endian platforms, but then I see in the generated EventPipeWriteEventMonoProfilerJitDone_V1 that on BIGENDIAN platforms, we swap one more time, wouldn't that eliminate the orignal swap done in ep_write_buffer_uint64 and put it back to BIGENDIAN byte order and not what anticipated in the final serialized stream?

Right, I wasn't aware that meanwhile, above FireETW layer, byte swaps are performed. Anyhow, I really like that which saves use a second walk over a struct/array during FireETW. AFAICS the following events make use of structs

```
BulkType
MonoProfilerJitDone_V1
MonoProfilerClassLoaded_V1
MonoProfilerGCMoves
MonoProfilerGCRoots
MonoProfilerGCHeapDumpObjectReference
```

and the following make use of arrays

```
MethodDCEndILToNativeMap
MethodDetails
MethodILToNativeMap
```

I removed the array and struct walk during FireETW and added appropriate `ep_write_buffer_x` calls.

lateralusX commented:
Will revisit this during the week, but I think the implemented approach is much simpler and more inline with how things already worked for "complex" data types.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3172,26 +3174,36 @@ ep_rt_mono_fire_bulk_type_event (BulkTypeEventLogger *type_logger)
+			+ sizeof (target->fixed_sized_data.cor_element_type)
+			+ num_bytes_utf16_str
+			+ sizeof (target->type_parameters_count)
+			+ (target->type_parameters_count * 8));
```
nit: sizeof (uint64_t)

lateralusX replied:
or even better, do change as outlined below, have  `ep_write_buffer_` return written byte size.

stefan-sf-ibm replied:
I very much like the idea and changed the `ep_writer_buffer_` functions to return the written number of bytes.

lateralusX commented on src/native/eventpipe/ep.h:
```diff
@@ -320,6 +320,28 @@ EP_WRITE_BUFFER_INT (64, u)
+size_t
+ep_write_buffer_string_utf8_to_utf16_t (uint8_t **buf, const ep_char8_t *str, size_t len)
+{
+	ep_char16_t *str_utf16 = ep_rt_utf8_to_utf16le_string (str, len);
```
nit: Optimize this for str == NULL or len == 0? It is the current use case for bulk event so it could be called rather frequently called that way.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3172,26 +3174,36 @@ ep_rt_mono_fire_bulk_type_event (BulkTypeEventLogger *type_logger)
+			ep_write_buffer_uint64_t (&ptr, type_parameter);
 		}
+
+		values_element_size += (uint32_t)(
```
Maybe it would be an idea to modify ep_write_buffer_ to return the number of bytes written, then we don't need to do this double book keeping making sure all matches above, I think that would greatly improve the maintainability of this code.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -5521,14 +5534,12 @@ mono_profiler_fire_buffered_gc_event_roots (
 			objects++;
 
 			// GCRoots.Values[].AddressID.
 			address_id = (uintptr_t)*objects;
```
Heads-up, there is a fix on this line in main, so maybe you should rebase to make sure that is correctly picked up. It should use address instead of objects.

lateralusX commented on src/native/eventpipe/ep.h:
```diff
@@ -320,6 +320,28 @@ EP_WRITE_BUFFER_INT (64, u)
 
 #undef EP_WRITE_BUFFER_INT
 
+#if SIZEOF_VOID_P == 4
```
Couldn't this be a ep_write_buffer_uintptr_t setup similar as the macros, we already have a ep_rt_val_uintptr_t where we take care of of size, so better to just have this in one place than spread out, something like (or fix macro to be a little more flexible):

```
static
inline
void
ep_write_buffer_uintptr_t (uint8_t **buffer, uintptr_t value)
{
	value = ep_rt_val_uinptr_t (value);
	memcpy (*buffer, &value, sizeof (value));
	*buffer += sizeof (value);
}
```

lateralusX reviewed (commented):
I like the way this PR is moving, much cleaner and easier to follow. I had a couple of "last" comments that might be worth to look over. One that is done, PR has been rebase against main and we have pass on CI, I think this is ready to go. Thanks for adding this support into EventPipe, much appreciated!

stefan-sf-ibm commented:
Beside the resolved comments I also cleaned up `structTypes` from `genEventing.py` which I introduced previously and which is unused, now.

I slightly changed the interface of `ep_write_buffer_string_utf16_t` which I didn't like before. Now, the length parameter describes the string length and not the number of bytes which is probably more intuitive.

stefan-sf-ibm commented:
> Build error:
> 
> D:\a_work\1\s\src\native\eventpipe\ep-config.c(405): error C2220: the following warning is treated as an error D:\a_work\1\s\src\native\eventpipe\ep-config.c(405): warning C4267: '+=': conversion from 'size_t' to 'uint32_t', possible loss of data D:\a_work\1\s\src\native\eventpipe\ep-config.c(407): warning C4267: '+=': conversion from 'size_t' to 'uint32_t', possible loss of data [232/2384] Building CXX object nativeaot\Runtime\Full\CMakeFiles\Runtime.WorkstationGC.dir__\Crst.cpp.obj

Argh that slipped through my local build. From an EventPipe point of view lengths are always of type `uint32_t`. Though, strictly speaking `sizeof` is of type `size_t` but AFAICS modern GCC/Clang do not emit a warning if the constant can be represented by a smaller type. Therefore, it might be ok to use `uint32_t` as a return type for all `ep_write_buffer_` functions. In the end I would like to have a solution with a few casts only. If we use `uint32_t` as the return type, then for the string functions the length parameter should be also of type `uint32_t` which may cause a cast if the parameter stems from `strlen`. On the other hand if we leave the return type of `size_t`, then at least once we hand over to the FireETW layer we have to do a cast like:

```
diff --git a/src/native/eventpipe/ep-config.c b/src/native/eventpipe/ep-config.c
index 3b8798a384f..85277f451d6 100644
--- a/src/native/eventpipe/ep-config.c
+++ b/src/native/eventpipe/ep-config.c
@@ -392,7 +392,7 @@ ep_config_build_event_metadata_event (
        const uint8_t *payload_data = ep_event_get_metadata (source_event);
        uint32_t payload_data_len = ep_event_get_metadata_len (source_event);
        uint32_t provider_name_len = (uint32_t)ep_rt_utf16_string_len (provider_name_utf16);
-       uint32_t instance_payload_size = 0;
+       size_t instance_payload_size = 0;
 
        // Allocate the payload.
        instance_payload = ep_rt_byte_array_alloc (instance_payload_size);
@@ -416,7 +416,7 @@ ep_config_build_event_metadata_event (
                ep_rt_current_processor_get_number (),
                ep_rt_thread_id_t_to_uint64_t (ep_rt_current_thread_get_id ()),
                instance_payload,
-               instance_payload_size,
+               (uint32_t)instance_payload_size,
                NULL /* pActivityId */,
                NULL /* pRelatedActivityId */);
```

Any preferences whether `size_t` or `uint32_t` should be used?

lateralusX commented:
> > Build error:
> > D:\a_work\1\s\src\native\eventpipe\ep-config.c(405): error C2220: the following warning is treated as an error D:\a_work\1\s\src\native\eventpipe\ep-config.c(405): warning C4267: '+=': conversion from 'size_t' to 'uint32_t', possible loss of data D:\a_work\1\s\src\native\eventpipe\ep-config.c(407): warning C4267: '+=': conversion from 'size_t' to 'uint32_t', possible loss of data [232/2384] Building CXX object nativeaot\Runtime\Full\CMakeFiles\Runtime.WorkstationGC.dir__\Crst.cpp.obj
> 
> Argh that slipped through my local build. From an EventPipe point of view lengths are always of type `uint32_t`. Though, strictly speaking `sizeof` is of type `size_t` but AFAICS modern GCC/Clang do not emit a warning if the constant can be represented by a smaller type. Therefore, it might be ok to use `uint32_t` as a return type for all `ep_write_buffer_` functions. In the end I would like to have a solution with a few casts only. If we use `uint32_t` as the return type, then for the string functions the length parameter should be also of type `uint32_t` which may cause a cast if the parameter stems from `strlen`. On the other hand if we leave the return type of `size_t`, then at least once we hand over to the FireETW layer we have to do a cast like:
> 
> ```
> diff --git a/src/native/eventpipe/ep-config.c b/src/native/eventpipe/ep-config.c
> index 3b8798a384f..85277f451d6 100644
> --- a/src/native/eventpipe/ep-config.c
> +++ b/src/native/eventpipe/ep-config.c
> @@ -392,7 +392,7 @@ ep_config_build_event_metadata_event (
>         const uint8_t *payload_data = ep_event_get_metadata (source_event);
>         uint32_t payload_data_len = ep_event_get_metadata_len (source_event);
>         uint32_t provider_name_len = (uint32_t)ep_rt_utf16_string_len (provider_name_utf16);
> -       uint32_t instance_payload_size = 0;
> +       size_t instance_payload_size = 0;
>  
>         // Allocate the payload.
>         instance_payload = ep_rt_byte_array_alloc (instance_payload_size);
> @@ -416,7 +416,7 @@ ep_config_build_event_metadata_event (
>                 ep_rt_current_processor_get_number (),
>                 ep_rt_thread_id_t_to_uint64_t (ep_rt_current_thread_get_id ()),
>                 instance_payload,
> -               instance_payload_size,
> +               (uint32_t)instance_payload_size,
>                 NULL /* pActivityId */,
>                 NULL /* pRelatedActivityId */);
> ```
> 
> Any preferences whether `size_t` or `uint32_t` should be used?

@stefan-sf-ibm I think we should go with untin32_t as return for the ep_write_buffer_ and ep_write_buffer_string_utf8_to_utf16_t and to be consistent change the len parameter of ep_write_buffer_string_utf8_to_utf16_t to uint32_t as well.

stefan-sf-ibm commented:
While skimming over the code another candidate may be the return value of `ep_rt_utf16_string_len`
https://github.com/dotnet/runtime/blob/b90c9cf68b8849a1090f64186033ac99d234c8c2/src/native/eventpipe/ep-rt.h#L533-L535

lateralusX commented:
> ep_rt_utf16_string_len

I think we should keep it as is, its used in several places as size_t, so we keep the uint32_t at the serialization level, at least for now.

lateralusX reviewed (approved):
LGTM! Thanks for the contribution!

I think these changes makes the serialization more visible and clear, increasing chances that new events gets correctly serialized handling big endian platforms. I will definitely keep an eye open on future changes when new events are added or serialization logic gets changed.

stefan-sf-ibm commented:
Sounds great to me. Thanks for the detailed review, much appreciated!

lateralusX commented:
@stefan-sf-ibm Errors are related, crashes in different EventPipe tests, both CoreCLR and Mono.

lateralusX commented on src/native/eventpipe/ep-config.c:
```diff
@@ -391,8 +391,8 @@ ep_config_build_event_metadata_event (
 	uint32_t payload_data_len = ep_event_get_metadata_len (source_event);
-	uint32_t provider_name_len = (uint32_t)((ep_rt_utf16_string_len (provider_name_utf16) + 1) * sizeof (ep_char16_t));
-	uint32_t instance_payload_size = sizeof (metadata_id) + provider_name_len + payload_data_len;
+	uint32_t provider_name_len = (uint32_t)ep_rt_utf16_string_len (provider_name_utf16);
```
Are we missing NULL terminating character here? We had that in the previous used length.

stefan-sf-ibm replied:
Right, the length does not contain the null termination character anymore since I changed `ep_write_buffer_string_utf16_t` where the length parameter should describe the string length. I adapted `instance_payload_size` instead, now.

lateralusX commented on src/native/eventpipe/ep-config.c:
```diff
@@ -391,8 +391,8 @@ ep_config_build_event_metadata_event (
+	uint32_t instance_payload_size = 0;
 
 	// Allocate the payload.
 	instance_payload = ep_rt_byte_array_alloc (instance_payload_size);
```
This is not yet calculated, it was done before in the old code, still need to be done before, since you need the buffer to serialize stuff into.

stefan-sf-ibm replied:
Thought I could refactor this quickly so that we don't need to book-keep the length. Of course, this doesn't work out since we need the size a priori. Fixed in the latest push.

stefan-sf-ibm committed:
[eventpipe] Perform endian aware serialization of structs and arrays above FireETW layer

lateralusX reviewed (approved):
LGTM!
